Operator with a dependency fails to upgrade when a new parameter is introduced

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
If the matter is security sensitive, please disclose it privately to a security contact: https://github.com/kudobuilder/kudo/blob/main/SECURITY_CONTACTS
-->


**What happened**:  I developed [a simple operator](https://github.com/alembiewski/operator-with-dependencies) with a single dependency with the following structure:
```
operator-with-dependencies
├── dependency
│   ├── operator.yaml
│   ├── params.yaml
│   └── templates
├── operator.yaml
├── params.yaml
└── templates
    └── dependency-params.yaml
```
Dependency operator exposes one single parameter, that could be updated via parent instance. Both operators have `deploy`, `update` and `upgrade` plans defined, making it possible to implement a custom operational behavior for each operation. 
The Initial deployment goes well:
```
➜  k kudo install ./operator-with-dependencies
operator test/operator created
operator test/dependency created
operatorversion test/dependency-1.0.0-0.0.1 created
operatorversion test/operator-1.0.0-0.0.1 created
instance test/operator-instance created
➜  k get instance
NAME                           AGE
operator-instance              14s
operator-instance-dependency   14s
➜  k kudo plan status --instance operator-instance
Plan(s) for "operator-instance" in namespace "test":
.
└── operator-instance (Operator-Version: "operator-1.0.0-0.0.1" Active-Plan: "deploy")
    ├── Plan deploy (serial strategy) [COMPLETE], last updated 2021-02-25 16:35:27
    │   └── Phase deploy-dependency (serial strategy) [COMPLETE]
    │       └── Step deploy-dependency [COMPLETE]
    ├── Plan update (serial strategy) [NOT ACTIVE]
    │   └── Phase update-dependency (serial strategy) [NOT ACTIVE]
    │       └── Step update-dependency [NOT ACTIVE]
    └── Plan upgrade (serial strategy) [NOT ACTIVE]
        └── Phase upgrade-dependency (serial strategy) [NOT ACTIVE]
            └── Step upgrade-dependency [NOT ACTIVE]

➜  k kudo plan status --instance operator-instance-dependency
Plan(s) for "operator-instance-dependency" in namespace "test":
.
└── operator-instance-dependency (Operator-Version: "dependency-1.0.0-0.0.1" Active-Plan: "deploy")
    ├── Plan deploy (serial strategy) [COMPLETE], last updated 2021-02-25 16:35:27
    │   └── Phase dummy (serial strategy) [COMPLETE]
    │       └── Step dummy [COMPLETE]
    ├── Plan update (serial strategy) [NOT ACTIVE]
    │   └── Phase dummy (serial strategy) [NOT ACTIVE]
    │       └── Step dummy [NOT ACTIVE]
    └── Plan upgrade (serial strategy) [NOT ACTIVE]
        └── Phase dummy (serial strategy) [NOT ACTIVE]
            └── Step dummy [NOT ACTIVE]
```
Then let's say I want to extend my dependency and add another parameter and make it configurable via parent operator:
https://github.com/alembiewski/operator-with-dependencies/commit/faa903742eaf1a1c76875973f37b0abb62087768

When I'm trying to do an upgrade, the plan fails with the following error:
```
➜  k kudo upgrade ./operator-with-dependencies --instance operator-instance
operatorversion test/dependency-1.0.0-0.0.2 created
operatorversion test/operator-1.0.0-0.0.2 created
instance test/operator-instance updated
➜  k kudo plan status --instance operator-instance
Plan(s) for "operator-instance" in namespace "test":
.
└── operator-instance (Operator-Version: "operator-1.0.0-0.0.2" Active-Plan: "upgrade")
    ├── Plan deploy (serial strategy) [NOT ACTIVE]
    │   └── Phase deploy-dependency (serial strategy) [NOT ACTIVE]
    │       └── Step deploy-dependency [NOT ACTIVE]
    ├── Plan update (serial strategy) [NOT ACTIVE]
    │   └── Phase update-dependency (serial strategy) [NOT ACTIVE]
    │       └── Step update-dependency [NOT ACTIVE]
    └── Plan upgrade (serial strategy) [IN_PROGRESS], last updated 2021-02-25 16:41:01
        └── Phase upgrade-dependency (serial strategy) [IN_PROGRESS]
            └── Step upgrade-dependency [ERROR] (A transient error when executing task upgrade.upgrade-dependency.upgrade-dependency.dependency. Will retry. admission webhook "instance-admission.kudo.dev" denied the request: failed to update Instance test/operator-instance-dependency: upgrade to new OperatorVersion {  dependency-1.0.0-0.0.2    } together with a parameter update triggering 'update' is not allowed)
``` 
**What you expected to happen**:
Since I'm not changing any parameter value, but only adding a new one, it's not clear why the `update` plan gets triggered for the dependency in this case. I would expect KUDO to only trigger an `upgrade` plan. And how the upgrade is going to work in case when the value of the existing parameters actually changes between the versions? What is the recomended way to perform upgrades in such cases? 
**How to reproduce it (as minimally and precisely as possible)**:
1. Checkout https://github.com/alembiewski/operator-with-dependencies
2. Install `0.0.1` version from https://github.com/alembiewski/operator-with-dependencies/commit/c84b6122d65ef28fca52c68f9b0da6223e9d5ce4
```
k kudo install ./operator-with-dependencies
```
3. Switch to the next version https://github.com/alembiewski/operator-with-dependencies/commit/faa903742eaf1a1c76875973f37b0abb62087768 and do the upgrade:
```
k kudo upgrade ./operator-with-dependencies --instance operator-instance
```
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Kudo version (use `kubectl kudo version`): 
```
KUDO Version: version.Info{GitVersion:"0.18.2", GitCommit:"43929950", BuildDate:"2021-01-20T08:26:44Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64", KubernetesClientVersion:"v0.19.2"}
```
- Operator: https://github.com/alembiewski/operator-with-dependencies
- operatorVersion:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/webhook/instance_admission.go
+++ b/pkg/webhook/instance_admission.go
@@ -217,7 +217,7 @@ func admitUpdate(old, new *kudoapi.Instance, ov, oldOv *kudoapi.OperatorVersion)
 		return nil, fmt.Errorf("plan %s does not exist", newPlan)
 	}
 
-	changedDefs, removedDefs, err := changedParameters(old.Spec.Parameters, new.Spec.Parameters, ov)
+	changedDefs, removedDefs, err := changedParameters(old.Spec.Parameters, new.Spec.Parameters, oldOv, ov)
 	if err != nil {
 		return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
 	}
@@ -229,8 +229,8 @@ func admitUpdate(old, new *kudoapi.Instance, ov, oldOv *kudoapi.OperatorVersion)
 		return nil, fmt.Errorf("failed to validate parameters for Instance %s/%s: %v", new.Namespace, new.Name, err)
 	}
 
-	parameterDefs := append(changedDefs, removedDefs...)
-	triggeredPlan, err := triggeredByParameterUpdate(parameterDefs, ov)
+	updatedParameterDefs := append(changedDefs, removedDefs...)
+	triggeredPlan, err := triggeredByParameterUpdate(updatedParameterDefs, ov)
 	if err != nil {
 		return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
 	}
@@ -365,12 +365,23 @@ func triggeredByParameterUpdate(params []kudoapi.Parameter, ov *kudoapi.Operator
 }
 
 // changedParameters returns a list of parameter definitions for params which value changed or that were added from old to new
-// This does *not* include parameters which *definition* has changed in an OV upgrade but where the value has not changed!
-func changedParameters(old, new map[string]string, newOv *kudoapi.OperatorVersion) ([]kudoapi.Parameter, []kudoapi.Parameter, error) {
-	changed, removed := kudoapi.RichParameterDiff(old, new)
-	changedDefs, err := kudoapi.GetParamDefinitions(changed, newOv)
+// This does *not* include:
+// - parameters which *definition* has changed in an OV upgrade but where the value has not changed
+// - parameters that have been added in an OV upgrade
+func changedParameters(old, new map[string]string, oldOv, newOv *kudoapi.OperatorVersion) ([]kudoapi.Parameter, []kudoapi.Parameter, error) {
+	changedOrAdded, removed := kudoapi.RichParameterDiff(old, new)
+	changedOrAddedDefs, err := kudoapi.GetParamDefinitions(changedOrAdded, newOv)
 	if err != nil {
 		return nil, nil, err
+	}
+
+	// Valid parameters not present in the old OV are not treated as changed.
+	changedDefs := []kudoapi.Parameter{}
+	for _, param := range changedOrAddedDefs {
+		param := param
+		if funk.Find(oldOv.Spec.Parameters, func(p kudoapi.Parameter) bool { return param.Name == p.Name }) != nil {
+			changedDefs = append(changedDefs, param)
+		}
 	}
 
 	// we ignore the error for missing OV parameter definitions for removed parameters. For once, this is a valid use-case when

--- a/test/e2e/update-with-upgrade/00-assert.yaml
+++ b/test/e2e/update-with-upgrade/00-assert.yaml
@@ -1,0 +1,7 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  origin: deploy
+  foo: bar

--- a/test/e2e/update-with-upgrade/00-install.yaml
+++ b/test/e2e/update-with-upgrade/00-install.yaml
@@ -1,0 +1,5 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+commands:
+  - command: kubectl kudo install --instance update-with-upgrade ./update-with-upgrade-v1
+    namespaced: true

--- a/test/e2e/update-with-upgrade/01-assert.yaml
+++ b/test/e2e/update-with-upgrade/01-assert.yaml
@@ -1,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  origin: upgrade
+  foo: bar
+  bar: baz

--- a/test/e2e/update-with-upgrade/01-upgrade-operator-version.yaml
+++ b/test/e2e/update-with-upgrade/01-upgrade-operator-version.yaml
@@ -1,0 +1,5 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+commands:
+  - command: kubectl kudo upgrade --instance update-with-upgrade ./update-with-upgrade-v2 -p bar=baz
+    namespaced: true

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v1/operator.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v1/operator.yaml
@@ -1,0 +1,50 @@
+apiVersion: kudo.dev/v1beta1
+name: "update-with-upgrade"
+operatorVersion: "0.1.0"
+kubernetesVersion: 1.13.0
+tasks:
+  - name: deploy-configmap
+    kind: Apply
+    spec:
+      resources:
+        - configmap-deployed.yaml
+  - name: update-configmap
+    kind: Apply
+    spec:
+      resources:
+        - configmap-updated.yaml
+  - name: upgrade-configmap
+    kind: Apply
+    spec:
+      resources:
+        - configmap-upgraded.yaml
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-configmap
+        strategy: serial
+        steps:
+          - name: deploy-configmap
+            tasks:
+              - deploy-configmap
+  # This should not occur when upgrading the operator:
+  update:
+    strategy: serial
+    phases:
+      - name: update-configmap
+        strategy: serial
+        steps:
+          - name: update-configmap
+            tasks:
+              - update-configmap
+  # Instead, this should occur when upgrading the operator:
+  upgrade:
+    strategy: serial
+    phases:
+      - name: upgrade-configmap
+        strategy: serial
+        steps:
+          - name: upgrade-configmap
+            tasks:
+              - upgrade-configmap

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v1/params.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v1/params.yaml
@@ -1,0 +1,5 @@
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: foo
+    required: true
+    default: "bar"

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v1/templates/configmap-deployed.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v1/templates/configmap-deployed.yaml
@@ -1,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  # This field is used to determine the plan which created the config map
+  origin: deploy
+  # This field introduces a dependency on a parameter
+  foo: "{{ .Params.foo }}"

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v1/templates/configmap-updated.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v1/templates/configmap-updated.yaml
@@ -1,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  # This field is used to determine the plan which created the config map
+  origin: update
+  # This field introduces a dependency on a parameter
+  foo: "{{ .Params.foo }}"

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v1/templates/configmap-upgraded.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v1/templates/configmap-upgraded.yaml
@@ -1,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  # This field is used to determine the plan which created the config map
+  origin: upgrade
+  # This field introduces a dependency on a parameter
+  foo: "{{ .Params.foo }}"

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v2/operator.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v2/operator.yaml
@@ -1,0 +1,50 @@
+apiVersion: kudo.dev/v1beta1
+name: "update-with-upgrade"
+operatorVersion: "0.2.0"
+kubernetesVersion: 1.13.0
+tasks:
+  - name: deploy-configmap
+    kind: Apply
+    spec:
+      resources:
+        - configmap-deployed.yaml
+  - name: update-configmap
+    kind: Apply
+    spec:
+      resources:
+        - configmap-updated.yaml
+  - name: upgrade-configmap
+    kind: Apply
+    spec:
+      resources:
+        - configmap-upgraded.yaml
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-configmap
+        strategy: serial
+        steps:
+          - name: deploy-configmap
+            tasks:
+              - deploy-configmap
+  # This should not occur when upgrading the operator:
+  update:
+    strategy: serial
+    phases:
+      - name: update-configmap
+        strategy: serial
+        steps:
+          - name: update-configmap
+            tasks:
+              - update-configmap
+  # Instead, this should occur when upgrading the operator:
+  upgrade:
+    strategy: serial
+    phases:
+      - name: upgrade-configmap
+        strategy: serial
+        steps:
+          - name: upgrade-configmap
+            tasks:
+              - upgrade-configmap

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v2/params.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v2/params.yaml
@@ -1,0 +1,9 @@
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: foo
+    required: true
+    default: "bar"
+  # NOTE: "bar" intentionally specifies no default,
+  # so that upgrade is impossible without setting the value.
+  - name: bar
+    required: true

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v2/templates/configmap-deployed.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v2/templates/configmap-deployed.yaml
@@ -1,0 +1,10 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  # This field is used to determine the plan which created the config map
+  origin: deploy
+  # This field introduces a dependency on a parameter
+  foo: "{{ .Params.foo }}"
+  bar: "{{ .Params.bar }}"

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v2/templates/configmap-updated.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v2/templates/configmap-updated.yaml
@@ -1,0 +1,10 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  # This field is used to determine the plan which created the config map
+  origin: update
+  # This field introduces a dependency on a parameter
+  foo: "{{ .Params.foo }}"
+  bar: "{{ .Params.bar }}"

--- a/test/e2e/update-with-upgrade/update-with-upgrade-v2/templates/configmap-upgraded.yaml
+++ b/test/e2e/update-with-upgrade/update-with-upgrade-v2/templates/configmap-upgraded.yaml
@@ -1,0 +1,10 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  # This field is used to determine the plan which created the config map
+  origin: upgrade
+  # This field introduces a dependency on a parameter
+  foo: "{{ .Params.foo }}"
+  bar: "{{ .Params.bar }}"

